File: war_regions/entities.py
"""Game entities: players, buildings and vehicles.

Pure data + small helpers; all rule logic lives in :mod:`war_regions.game`.
"""

from dataclasses import dataclass
from enum import Enum

from . import constants as C
from . import hexgrid
from .constants import TurretKind, VehicleKind


class BuildingKind(Enum):
    """All building kinds (rules.md sec. 3)."""

    BASE_TANK = "base_tank"
    BASE_HELICOPTER = "base_helicopter"
    BASE_HOVERCRAFT = "base_hovercraft"
    BASE_BUFFER = "base_buffer"
    TURRET_NORMAL = "turret_normal"
    TURRET_RAPID = "turret_rapid"
    TURRET_ROCKET = "turret_rocket"
    HEAL_TOWER = "heal_tower"


#: Bases spawn the vehicle kind matching their type (rules.md sec. 4).
BASE_VEHICLE = {
    BuildingKind.BASE_TANK: VehicleKind.TANK,
    BuildingKind.BASE_HELICOPTER: VehicleKind.HELICOPTER,
    BuildingKind.BASE_HOVERCRAFT: VehicleKind.HOVERCRAFT,
    BuildingKind.BASE_BUFFER: VehicleKind.BUFFER,
}

#: Turret kind behind each turret building (rules.md sec. 10).
TURRET_BUILDING = {
    BuildingKind.TURRET_NORMAL: TurretKind.NORMAL,
    BuildingKind.TURRET_RAPID: TurretKind.RAPID,
    BuildingKind.TURRET_ROCKET: TurretKind.ROCKET,
}


def is_base(kind: BuildingKind) -> bool:
    """True for the four base kinds."""
    return kind in BASE_VEHICLE


def is_turret(kind: BuildingKind) -> bool:
    """True for the three turret kinds."""
    return kind in TURRET_BUILDING


def vehicle_kind_of(kind: BuildingKind) -> VehicleKind:
    """Vehicle kind spawned by a building: base type or a tank (sec. 4)."""
    return BASE_VEHICLE.get(kind, VehicleKind.TANK)


def turret_kind_of(kind: BuildingKind):
    """TurretKind of a turret building, else ``None``."""
    return TURRET_BUILDING.get(kind)


def capacity_of(kind: BuildingKind) -> int:
    """Building capacity: bases 100, everything else 50 (sec. 3)."""
    return C.BASE_CAPACITY if is_base(kind) else C.BUILDING_CAPACITY


@dataclass
class Player:
    """One of the 2-4 players (exactly one is human, rules.md sec. 1)."""

    id: int
    is_human: bool
    eliminated: bool = False

    @property
    def color(self) -> tuple:
        """Player colour used by the renderer."""
        return C.PLAYER_COLORS[self.id % len(C.PLAYER_COLORS)]


class Building:
    """A structure standing on one tile: base, turret or healing tower."""

    def __init__(self, kind: BuildingKind, owner, q: int, r: int,
                 units: float = 0.0):
        self.kind = kind
        self.owner = owner            # player id or None (neutral)
        self.tile = (q, r)
        self.units = float(units)
        self.capacity = capacity_of(kind)
        self.production_timer = 0.0   # base spawn cycle progress
        self.fire_timer = 0.0         # turret cooldown progress
        self.heal_timer = 0.0         # healing-tower pulse progress
        self.last_target_pos = None   # world pos of last turret target
        # floating -x / +x aggregation (specification: floating numbers)
        self.loss_acc = 0.0
        self.gain_acc = 0.0
        self.text_timer = 0.0
        self.texts = []               # [ [amount, age], ... ]

    @property
    def pos(self) -> tuple:
        """World position of the tile centre."""
        return hexgrid.hex_to_world(self.tile[0], self.tile[1], C.HEX_SIDE)

    def __repr__(self):  # pragma: no cover - debugging aid
        return f"<Building {self.kind.value} owner={self.owner} " \
               f"tile={self.tile} units={self.units:.0f}>"


class Vehicle:
    """A moving group of units travelling along a fixed route (sec. 4)."""

    _next_id = 1

    def __init__(self, kind: VehicleKind, owner: int, units: float,
                 route: list, start_pos: tuple):
        self.id = Vehicle._next_id
        Vehicle._next_id += 1
        self.kind = kind
        self.owner = owner
        self.units = float(units)
        self.route = list(route)          # tiles after source, incl. target
        self.route_index = 0              # next waypoint into ``route``
        self.x, self.y = start_pos        # continuous world position
        self.combat_target = None         # Vehicle | None (sec. 9)
        self.last_opponent = None         # last vehicle we shot at
        self.fire_timer = 0.0             # combat shot cooldown progress
        self.wall_timer = 0.0             # wall-attack shot cooldown progress
        self.wall_target = None           # tile of the wall being shot
        self.wall_shots = 0               # shots fired at that wall so far
        self.loss_acc = 0.0
        self.gain_acc = 0.0
        self.text_timer = 0.0
        self.texts = []
        self.dead = False                 # set when destroyed / arrived

    @property
    def pos(self) -> tuple:
        """World position (centre of the sprite)."""
        return (self.x, self.y)

    @property
    def dest_tile(self):
        """Final tile of the route (or ``None``)."""
        return self.route[-1] if self.route else None

    def __repr__(self):  # pragma: no cover - debugging aid
        return f"<Vehicle {self.kind.value} owner={self.owner} " \
               f"units={self.units:.0f} id={self.id}>"

File: rust/src/entities.rs
//! Game entities: players, buildings and vehicles.
//!
//! Pure data plus small helpers; all rule logic lives in [`crate::game`].

use crate::constants::{self, VehicleKind};
use crate::hexgrid::{self, Tile};

/// All building kinds (rules.md section 3).
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum BuildingKind {
    /// Tank-producing base.
    BaseTank,
    /// Helicopter-producing base.
    BaseHelicopter,
    /// Hovercraft-producing base.
    BaseHovercraft,
    /// Buffer-producing base.
    BaseBuffer,
    /// Ordinary turret building.
    TurretNormal,
    /// Rapid-fire turret building.
    TurretRapid,
    /// Rocket turret building.
    TurretRocket,
    /// Healing tower.
    HealTower,
}

/// True for the four base kinds.
pub fn is_base(kind: BuildingKind) -> bool {
    matches!(
        kind,
        BuildingKind::BaseTank
            | BuildingKind::BaseHelicopter
            | BuildingKind::BaseHovercraft
            | BuildingKind::BaseBuffer
    )
}

/// True for the three turret kinds.
pub fn is_turret(kind: BuildingKind) -> bool {
    matches!(
        kind,
        BuildingKind::TurretNormal | BuildingKind::TurretRapid | BuildingKind::TurretRocket
    )
}

/// Vehicle kind spawned by a building: base type or a tank (section 4).
pub fn vehicle_kind_of(kind: BuildingKind) -> VehicleKind {
    match kind {
        BuildingKind::BaseTank => VehicleKind::Tank,
        BuildingKind::BaseHelicopter => VehicleKind::Helicopter,
        BuildingKind::BaseHovercraft => VehicleKind::Hovercraft,
        BuildingKind::BaseBuffer => VehicleKind::Buffer,
        _ => VehicleKind::Tank,
    }
}

/// Turret kind of a turret building, else `None`.
pub fn turret_kind_of(kind: BuildingKind) -> Option<crate::constants::TurretKind> {
    match kind {
        BuildingKind::TurretNormal => Some(crate::constants::TurretKind::Normal),
        BuildingKind::TurretRapid => Some(crate::constants::TurretKind::Rapid),
        BuildingKind::TurretRocket => Some(crate::constants::TurretKind::Rocket),
        _ => None,
    }
}

/// Building capacity: bases 100, everything else 50 (section 3).
pub fn capacity_of(kind: BuildingKind) -> f64 {
    if is_base(kind) {
        constants::BASE_CAPACITY
    } else {
        constants::BUILDING_CAPACITY
    }
}

/// One of the 2-4 players (exactly one is human, rules.md section 1).
#[derive(Clone, Debug)]
pub struct Player {
    /// Player id (0 is the human when present).
    pub id: usize,
    /// True for the human-controlled player.
    pub is_human: bool,
    /// True once the player lost everything (rules.md section 2).
    pub eliminated: bool,
}

impl Player {
    /// Create a player with `id`; `is_human` marks the human player.
    pub fn new(id: usize, is_human: bool) -> Self {
        Self {
            id,
            is_human,
            eliminated: false,
        }
    }
    #[allow(dead_code)]
    /// Player colour used by the renderer.
    pub fn color(&self) -> [u8; 3] {
        constants::player_color(self.id)
    }
}

/// A floating combat number: `(amount, age)`.
#[derive(Clone, Copy, Debug)]
pub struct FloatText {
    /// Displayed amount (negative = loss, positive = gain).
    pub amount: f64,
    /// Seconds since the text appeared.
    pub age: f64,
}

/// A structure standing on one tile: base, turret or healing tower.
#[derive(Clone, Debug)]
pub struct Building {
    /// Building kind (rules.md section 3).
    pub kind: BuildingKind,
    /// Owning player id, or `None` when neutral.
    pub owner: Option<usize>,
    /// Tile the building stands on.
    pub tile: Tile,
    /// Units currently inside (may exceed capacity when overcrowded).
    pub units: f64,
    /// Maximum comfortable garrison (section 3).
    pub capacity: f64,
    /// Base spawn cycle progress.
    pub production_timer: f64,
    /// Turret cooldown progress.
    pub fire_timer: f64,
    /// Healing-tower pulse progress.
    pub heal_timer: f64,
    /// World position of the last turret target (for the barrel).
    pub last_target_pos: Option<(f64, f64)>,
    /// Accumulated fractional losses/gains (floating numbers).
    pub loss_acc: f64,
    /// Accumulated fractional gains (floating numbers).
    pub gain_acc: f64,
    /// Timer for committing fractional floating text.
    pub text_timer: f64,
    /// Visible floating texts.
    pub texts: Vec<FloatText>,
}

impl Building {
    /// Create a building of `kind` owned by `owner` on tile `(q, r)`.
    pub fn new(kind: BuildingKind, owner: Option<usize>, q: i32, r: i32, units: f64) -> Self {
        Self {
            kind,
            owner,
            tile: (q, r),
            units,
            capacity: capacity_of(kind),
            production_timer: 0.0,
            fire_timer: 0.0,
            heal_timer: 0.0,
            last_target_pos: None,
            loss_acc: 0.0,
            gain_acc: 0.0,
            text_timer: 0.0,
            texts: Vec::new(),
        }
    }
    /// World position of the tile centre.
    pub fn pos(&self, side: f64) -> (f64, f64) {
        hexgrid::hex_to_world(self.tile.0, self.tile.1, side)
    }
}

/// Next vehicle id (monotonic within one process).
static NEXT_VEHICLE_ID: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(1);

/// A moving group of units travelling along a fixed route (section 4).
#[derive(Clone, Debug)]
pub struct Vehicle {
    /// Unique id (debugging / projectile targeting).
    pub id: u64,
    /// Vehicle kind (rules.md section 4).
    pub kind: VehicleKind,
    /// Owning player id.
    pub owner: usize,
    /// Units carried.
    pub units: f64,
    /// Tiles after the source, including the target.
    pub route: Vec<Tile>,
    /// Index of the next waypoint into `route`.
    pub route_index: usize,
    /// Continuous world position (sprite centre).
    pub x: f64,
    /// Continuous world position (sprite centre).
    pub y: f64,
    /// Source tile the route starts from.
    pub src_tile: Option<Tile>,
    /// Combat opponent id (section 9).
    pub combat_target: Option<u64>,
    /// Id of the last vehicle shot at.
    pub last_opponent: Option<u64>,
    /// Combat shot cooldown progress.
    pub fire_timer: f64,
    /// Wall-attack shot cooldown progress.
    pub wall_timer: f64,
    /// Tile of the wall being shot.
    pub wall_target: Option<Tile>,
    /// Shots fired at that wall so far.
    pub wall_shots: u32,
    /// Accumulated fractional losses/gains.
    pub loss_acc: f64,
    /// Accumulated fractional gains.
    pub gain_acc: f64,
    /// Timer for committing fractional floating text.
    pub text_timer: f64,
    /// Visible floating texts.
    pub texts: Vec<FloatText>,
    /// Set when destroyed or arrived.
    pub dead: bool,
}

impl Vehicle {
    /// Create a vehicle starting at `start_pos` and following `route`.
    pub fn new(
        kind: VehicleKind,
        owner: usize,
        units: f64,
        route: Vec<Tile>,
        start_pos: (f64, f64),
        src_tile: Option<Tile>,
    ) -> Self {
        let id = NEXT_VEHICLE_ID.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
        Self {
            id,
            kind,
            owner,
            units,
            route,
            route_index: 0,
            x: start_pos.0,
            y: start_pos.1,
            src_tile,
            combat_target: None,
            last_opponent: None,
            fire_timer: 0.0,
            wall_timer: 0.0,
            wall_target: None,
            wall_shots: 0,
            loss_acc: 0.0,
            gain_acc: 0.0,
            text_timer: 0.0,
            texts: Vec::new(),
            dead: false,
        }
    }
    /// World position (centre of the sprite).
    pub fn pos(&self) -> (f64, f64) {
        (self.x, self.y)
    }
    #[allow(dead_code)]
    /// Final tile of the route (or `None`).
    pub fn dest_tile(&self) -> Option<Tile> {
        self.route.last().copied()
    }
}
